docs/proposal-007-apply.md

```diff
@@ -0,0 +1,173 @@
+of an EKS cluster with the intended state as specified in the `eksctl` config file.
+
+This proposal is only concerned with the `apply` subcommand. Existing behavior
+and commands are unaffected. Continuing to support them shouldn't complicate
+the implementation too much as they are more or less subsets of
+operations performed by `apply`.
+
+## Motivation
+
+`eksctl` supplies users with imperative commands for clusters and cluster
+resources that allow them to take explicit actions, like `create` or `delete`.
+We also encourage the use of a config file in which the desired state of the
+cluster is described. Still, users are required to figure out which imperative
+steps to take to reconcile the cluster with the desired state.
+The missing capability is to give `eksctl` the ability to do this.
+Conceptually, `eksctl` would gather all of the information it can about the
+current state of the cluster in order to build a `ClusterConfig` describing
+the real-world cluster and then `diff` this config with the user-provided
+`ClusterConfig`. The `diff` is used to setup a plan of modification to the
+cluster and resources.
+
+### Questions for discussion
+
+This proposal also opens the floor for discussion around non-straightforward
+questions around reconciliation.
+
+#### Deletion of resources
+
+When looking at a list of nodegroups in a cluster, it's not clear how to handle
+deletion.
+
+-   Is the given list intended to be a complete list of nodegroups and
+    no others should exist? Are they automatically deleted?
+-   How do we handle non-eksctl managed resources?
+
+#### Immutability
+
+In order to change some properties, entire resource will have to be recreated.
+This should be explicit to the user.
+
+#### Gradual support and config
+
+Because it's impractical to introduce complete support for reconciliation at
+once, we need to instead gradually support properties and resources of a
+cluster.
+
+This leads to the question of whether the current `ClusterConfig` is a good
+structure for a config meant for reconciliation.
+
+## Proposal
+
+Introduce a subcommand `apply` which has only a config file argument.
+`apply` would reconcile desired with real-world state
+[as described in the motivation](#motivation).
+The exact details per resource will potentially differ depending on
+the resources and won't be enumerated here.
+
+1. `apply` is put behind the explicit "experimental" flag (cf `enable profile/repo`)
+2. Take full ownership of `eksctl`-managed resources (i.e. delete missing resources by default)
+3. Options for non-`eksctl`-managed resources as well as resources where
+   ownership is unknowable are covered below
+4. Changing immutable fields through recreation is the default
+5. We gradually expand `apply` support to different parts of the cluster
+6. As we expand `apply` support, we reevaluate and update the config structure as
+   necessary in a new API version.
```

Comment:
I think it would be good to spec out precisely what an entire cluster config looks like before we start. I feel one of the reasons the current config is incohesive is because it was patched together issue-by-issue

Reply:
Before we start implementing perhaps, but IMO that's really the next step from this proposal. IMO the main reason the config is incohesive right now is simply because we can't/don't make breaking changes. Perhaps we could sketch out a tentative structure for the entire thing in terms of "let's think about which resources we're managing here" so we can think through cases for `apply` but I don't think we should wait for the whole entire config to be supported before releasing some version of `apply` and I think we shouldn't pretend we'll get it right the first time.

Reply:
> we shouldn't pretend we'll get it right the first time.

👍 , totally. I was thinking more on "based on the resources we _know_ we're going to manage"